src/representor/hal.ts

```diff
@@ -2,75 +2,141 @@ import Link from '../link';
 import { resolve } from '../utils/url';
 import Representation from './base';
 
+type HalLink = {
+  href: string,
+  name?: string,
+  title: string,
+  templated?: boolean,
+  type?: string
+};
+
+type HalBody = {
+  _links?: {
+    [rel: string]: HalLink | HalLink[],
+  }
+  [key: string]: any,
+  _embedded?: {
+    [rel: string]: HalBody | HalBody[],
+  }
+};
+
+/**
+ * Internal representation for embedded resources
+ */
+type Embedded = {
+  href: string,
+  contextUri: string,
+  rel: string,
+  body: HalBody
+};
+
 /**
  * The Representation class is basically a 'body' of a request
  * or response.
  *
  * This class is for HAL JSON responses.
  */
-export default class Hal extends Representation {
+export default class Hal extends Representation<HalBody> {
 
-  body: { [s: string]: any };
+  embedded: Embedded[];
 
-  constructor(uri: string, contentType: string, body: any) {
+  /**
+   * parse is called to convert a HTTP response body string into the most
+   * suitable internal body type.
+   *
+   * For JSON responses, usually this means calling JSON.parse() and returning
+   * the result.
+   */
+  parse(body: string): HalBody {
 
-    super(uri, contentType, body);
+    return JSON.parse(body);
 
-    if (typeof body === 'string') {
-      this.body = JSON.parse(body);
-    } else {
-      this.body = body;
-    }
+  }
 
-    if (typeof this.body._links !== 'undefined') {
-      parseHalLinks(this);
-    }
-    if (typeof this.body._embedded !== 'undefined') {
-      parseHalEmbedded(this);
-    }
+  parseLinks(body: HalBody): Link[] {
 
-    delete this.body._links;
-    delete this.body._embedded;
+    return parseHalLinks(this.uri, body);
 
   }
 
+  /**
+   * Returns the parsed body for this representation.
+   *
+   * Specific implementations of this class might alter the response before
+   * returning, for example to remove meta-information that's not relevant
+   * the user of this object.
+   */
+  getBody(): HalBody {
+
+    const newBody = Object.assign({}, this.body);
```

Comment:
You could use destructuring to leave the original alone: `const newBody = {_embedded, _links, ...this.body}`

Reply:
Changing this too in #147